modules/internal/ChapelDynamicLoading.chpl

```diff
@@ -0,0 +1,797 @@
+            }
+          }
+        }
+      }
+
+      return ret;
+    }
+
+    inline proc type _systemDynLibClearError() do dlerror();
+
+    inline proc type _systemDynLibClose(ptr: c_ptr(void)) do dlclose(ptr);
+
+    inline proc type _systemDynLibWrapError() {
+      const sysErr = dlerror();
+      if sysErr != nil then
+        return new DynLibError?(try! string.createCopyingBuffer(sysErr));
+      return nil;
+    }
+
+    inline proc type
+    _systemDynLibOpen(path: string, out err: owned DynLibError?) {
+      _systemDynLibClearError();
+
+      var ret = dlopen(path.c_str(), RTLD_LAZY);
+      if ret == nil {
+        var tmp = _systemDynLibWrapError();
+        if tmp != nil then err = tmp;
+      }
+
+      return ret;
+    }
+
+    inline proc type
+    _systemDynLibLookup(sym: string, handle: c_ptr(void),
+                        out err: owned DynLibError?) {
+      _systemDynLibClearError();
+
+      var ret = dlsym(handle, sym.c_str());
+      if ret == nil {
+        var tmp = _systemDynLibWrapError();
+        if tmp != nil then err = tmp;
+      }
+
+      return ret;
+    }
+
+    // TODO: Also need to evict pointer cache entries.
+    proc _close() {
+      assert(_refCount.read() == 0);
+      if _closed then return;
+
+      on Locales[0] do manage this.withWriteLock() {
+        if !_closed {
+          _closed = true;
+          for i in 0..<_systemPtrs.size {
+            const loc = Locales[i];
+            const ptr = _systemPtrs[i];
+            if ptr != nil then on loc {
+              this.type._systemDynLibClose(ptr);
+            }
+          }
+        }
+      }
+    }
+
+    inline proc bumpRefCount() do _refCount.fetchAdd(1);
+    inline proc dropRefCount() {
+      if _refCount.read() <= 0 then return;
+      _refCount.fetchSub(1);
+    }
+
+    // TODO: 'T' must be a procedure type, but we cannot restrict it yet.
+    // Could add a primitive "any procedure type" to use instead.
+    proc loadSymbol(sym: string, type t, out err: owned DynLibError?) {
+      var idx: int = 0;
+
+      if !isProcedure(t) || isClass(t) {
+        compilerError('The type passed to \'loadSymbol\' must be ' +
+                      'a procedure type');
+      }
+
+      if chpl_isLocalProc(t) {
+        // The 'wideness' of a procedure type should not necessarily be
+        // exposed to the user (and so we really shouldn't care), but it's an
+        // assumption we make about the input type as a precaution. Also,
+        // we might like to allow users to load references to data as well.
+        compilerError('The procedure type passed to \'loadSymbol\'' +
+                      'should be wide');
+      }
+
+      // TODO: I think a faster path would just try to load the symbol without
+      // going to LOCALE-0. The problem with that is that we wouldn't be able
+      // to do any book-keeping, or we'd need to make some portion of the
+      // state be privatized. Or, we could avoid book-keeping if we just say
```

Comment:
is there something missing here?

Reply:
Thanks, I've removed that comment and have a private TODO about reducing comm.